### PR TITLE
Don't Rewind on SPV Request Lookup

### PR DESCRIPTION
**Pull request by emilypi**

### CODE REVIEW

File: src/Chainweb/Pact/Types.hs [hunk 1]
```diff
@@ -105,4 +105,16 @@ newtype GasId = GasId PactId deriving (Eq, Show)
...
+    | NoRewind !a
+    | Genesis {-# UNPACK #-} !cid
+    deriving (Functor, Applicative, Foldable, Traversable, Monad)
+
+instance HasChainId a => HasChainId (Rewind a) where
+    _chainId = \case
+      DoRewind !a -> _chainId a
+      NoRewind !a -> _chainId a
+      Genesis !cid -> cid
+
```
Comment (emmanueldenloye) [inline src/Chainweb/Pact/Types.hs]: Why do you need Functor, Applicative, Foldable, Traversable and Monad instances?

Comment (emilypi) [reply src/Chainweb/Pact/Types.hs]: Cautionary - i don't know where I'll be going so I want to know i have everything I need and I'll pare it down as I work the PR. It's a WIP draft!

File: src/Chainweb/Chainweb.hs [hunk 2]
```diff
@@ -495,7 +496,7 @@ validatingMempoolConfig cid v gl mv = Mempool.InMemConfig
     preInsertBatch txs = do
         let hashes = V.map (toPactHash . sfst) txs
         pex <- readMVar mv
-        rs <- _pactLookup pex (Left cid) hashes >>= either throwM pure
+        rs <- _pactLookup pex (Genesis cid) hashes >>= either throwM pure
```
Comment (emilypi) [inline src/Chainweb/Chainweb.hs]: I'm not sure if "Genesis" is the correct name for this case when we're looking up by chain id. Paint my bikeshed please

File: src/Chainweb/Pact/PactService.hs [hunk 3]
```diff
@@ -1077,26 +1077,29 @@ transactionsFromPayload plData = do
...
+      case restorePoint of
+        NoRewind _ -> liftIO $!
+          V.mapM (_cpLookupProcessedTx cp) txs
+        DoRewind a -> doRewind cp a
+        _ -> do
+          mrp <- fmap (\(!bh) -> T2 (_blockHeight bh) (_blockHash bh))
+            <$!> findLatestValidBlock
+          case mrp of
+            Nothing -> return mempty -- nothing denotes genesis
+            Just a -> doRewind cp a
```
Comment (emilypi) [inline src/Chainweb/Pact/PactService.hs]: If `NoRewind`, lookup all txs without rewinding. This is only used when doing SPV lookups. When `DoRewind` or `Genesis`, do the rewind.

File: src/Chainweb/Pact/Types.hs [hunk 4]
```diff
@@ -105,4 +110,16 @@ newtype GasId = GasId PactId deriving (Eq, Show)
 
 type TransactionM p a = ReaderT (CommandEnv p) IO a
 
+data Rewind a
+    = DoRewind !a
```
Comment (gregorycollins) [inline src/Chainweb/Pact/Types.hs]: There should only be two cases here:

```
DoRewind (T2 BlockHeight BlockHash) | NoRewind ChainId
```

File: src/Chainweb/Pact/PactService.hs (hunk 3 above)
Comment (gregorycollins) [reply src/Chainweb/Pact/PactService.hs]: PreInsert shouldn't rewind, that's a regression -- everything that was `Left` before should be `NoRewind`.

File: src/Chainweb/Pact/Types.hs (hunk 4 above)
Comment (gregorycollins) [reply src/Chainweb/Pact/Types.hs]: Does `Rewind` need to be polymorphic?

Comment (emilypi) [reply src/Chainweb/Pact/Types.hs]: Made it monomorphic.

**Pull request merged**